### DISCUSSION TITLE
Run pyvista.Plotter in headless environment (mac, windows, and linux)

### DISCUSSION THREAD

**Discussion by isaaccorley** [pyvista/pyvista#3692] [Q&A] [answered]
Hi, I'm currently refactoring some code in [torchgeo](https://github.com/microsoft/torchgeo) to replace `open3d` with `pyvista`. We run tests using github actions with sample data to ensure that all of our plotting functionality works on multiple platforms (linux, windows, macos) without failing. I 
 see some previous dicussions for plotting in a headless CentOS environment, however, there doesn't seem to be an option I could find specific to windows and macos to run the plotting capabilities in a headless environment that won't throw an error. Is there an easy solution to this that I've just overlooked?

**Comment by banesullivan** [comment]
I'm so thrilled to hear you're using PyVista in torchgeo! It's an awesome project I've dabbled with a bit in the past.

> We run tests using github actions

If using GitHub Actions, we have this action to configure the headless display: https://github.com/pyvista/setup-headless-display-action (I currently need some help updating the OSMesa version for windows if you know anyone privy to it)

Otherwise, take a look at this section of the documentation: https://docs.pyvista.org/getting-started/installation.html#running-on-remote-servers

Using `xvfb` is a common solution and you'll need a startup script. Either `pv.start_xvfb()`  or something along the lines of:

```
sudo apt-get install xvfb
export DISPLAY=:99.0
export PYVISTA_OFF_SCREEN=true
export PYVISTA_USE_IPYVTK=true
Xvfb :99 -screen 0 1024x768x24 > /dev/null 2>&1 &
sleep 3
```


---- 

I also want to note that we at Kitware are actively working to start publishing OSMesa wheel variants for VTK (this should be out in the next few weeks at latest), which will prevent the need for frame buffers and any of the above. Simply `apt install` osmesa, use the OSMesa wheel variant, and voila.

@isaaccorley, feel free to point me to your CI actions and I'll see if I can provide input to get them up and running

**Reply by adamjstewart** [reply]
Our CI actions are [here](https://github.com/microsoft/torchgeo/blob/main/.github/workflows/tests.yaml).

**Reply by adamjstewart** [reply]
We could theoretically only run these plotting tests on Linux if that's the only system that's easy to test in CI.

**Reply by banesullivan** [reply]
You can run these tests on any OS and quite easily on GitHub Actions. This should be as simple as plugging the `setup-headless-display-action` action into your existing workflow. https://github.com/pyvista/setup-headless-display-action

I'm assuming this work is being done in https://github.com/microsoft/torchgeo/pull/663, is that correct? I'll try to make some suggestions there

**Reply by adamjstewart** [reply]
That would be perfect! Yes, that is the correct PR.

**Reply by bilke** [reply]
@banesullivan Dear Bane,

> I also want to note that we at Kitware are actively working to start publishing OSMesa wheel variants for VTK (this should be out in the next few weeks at latest), which will prevent the need for frame buffers and any of the above. Simply apt install osmesa, use the OSMesa wheel variant, and voila.

Is there are an issue on VTK I can track regarding the OSMesa wheels? Thanks a lot!

**Reply by banesullivan** [reply]
We're internally tracking this at the moment. I'll see if I can open something with external visibility. It should be done in the coming weeks and I'll make a note to myself to come back to this discussion at least

**Reply by banesullivan** [reply]
OSMesa wheels are finally available. We're workign to get a `9.2.5` version tagged wheel, but some are available for master and the release branch of VTK:

- `master`: https://gitlab.kitware.com/vtk/vtk/-/packages/90
- `9.2.5.dev*`: https://gitlab.kitware.com/vtk/vtk/-/packages/92

Also. please reference this discussion where I point out some other wheels and how to build these yourself (fairly painlessly): https://github.com/Kitware/trame/discussions/193#discussioncomment-4834848

**Reply by banesullivan** [reply, marked as answer]
```
pip install vtk_osmesa<=9.2.1000 --pre --no-cache --extra-index-url https://gitlab.kitware.com/api/v4/projects/13/packages/pypi/simple
```

**Reply by carbocation** [reply]
Double checking based on @banesullivan 's mentions of the vtk-osmesa wheel. If I'm using pyvista and I install vtk-osmesa, should I be able to do headless rendering or are there more steps that I need to do beyond `pip install vtk-osmesa --index-url https://gitlab.kitware.com/api/v4/projects/13/packages/pypi/simple`?

**Reply by carbocation** [reply]
(To close this out - simply installing `vtk-osmesa` was enough to make this work for me. Thanks!)

**Comment by adamjstewart** [comment]
@banesullivan do you have any suggestions for handling https://github.com/microsoft/torchgeo/issues/1651? We're looking for a way to confirm that our pyvista plots are working without actually opening up a plot and pausing the tests.

**Reply by banesullivan** [reply]
Responded on that issue. Feel welcome to open new issues here or ping us in `torchgeo` rather than keeping this discussion going
